### PR TITLE
Feature/detailed status information

### PR DESCRIPTION
**Pull request by lukasmweber**
Composition Status is now logged using 5 states:
* System Composition
* Synthesis
* Place
* Route
* Write Bitstream
Whenever a new stage starts, elapsed time of the prev. stage is shown as well as overall elapsed time

### CODE REVIEW

File: src/main/scala/tapasco/filemgmt/ProgressTrackingFileWatcher.scala [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    if(old != currentState) {
+      if(old != -1) {
+        val oldProgressionString = progressionStringsInfo(old)._2
+        logger.info("Finished %s after %s (Total Elapsed: %s)".format(oldProgressionString, timeString(stageStart), timeString(start)))
+      }
+      val progressionString = progressionStringsInfo(currentState)._2
+      logger.info("Started %s (Total Elapsed: %s)".format(progressionString, timeString(start)))
+      stageStart = System.currentTimeMillis()
+
+      //This ensures that this watcher will close all opened files if he reaches the final state.
```
Comment (sommerlukas) [inline src/main/scala/tapasco/filemgmt/ProgressTrackingFileWatcher.scala]: What happens if the watcher never reaches the final state, e.g. due to some error in one of the steps?

File: src/main/scala/tapasco/filemgmt/ProgressTrackingFileWatcher.scala [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+  /**
+    * Calculates the time since a certain TimeStamp and transforms it into a hh:mm:ss-Formatted String.
+    * @param since timestamp
+    * @return TimeString
+    */
+  private def timeString(since: Long): String = {
+    val now = System.currentTimeMillis()
+    val dur = Duration(now-since, "millis")
+    val s = dur.toSeconds.toInt
```
Comment (sommerlukas) [inline src/main/scala/tapasco/filemgmt/ProgressTrackingFileWatcher.scala]: Why do you first get the duration in milliseconds if you anyway convert it to seconds immediately after that? You could get the number of seconds with `"seconds"` as parameter to `Duration.apply`.

File: src/main/scala/tapasco/filemgmt/ProgressTrackingFileWatcher.scala [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    * @param string parsed String
+    */
+  private def checkState(string: String): Unit = {
+    val old = currentState
+    if(currentState == -1) {
+      currentState = currentState + 1
+    }
+    for(i <- progressionStringsInfo.indices) {
+      val progressionString = progressionStringsInfo(i)._1
+      if(i == -1 || (currentState == i - 1 && string.contains(progressionString))){
```
Comment (sommerlukas) [inline src/main/scala/tapasco/filemgmt/ProgressTrackingFileWatcher.scala]: Wouldn't you miss on some status information, if two stages are completed in a time interval shorter than the `POLL_INTERVAL`?

File: src/main/scala/tapasco/activity/composers/VivadoComposer.scala [hunk 4]
```diff
@@ -22,20 +22,20 @@
...
+import de.tu_darmstadt.cs.esa.tapasco.filemgmt.{FileAssetManager, LogTrackingFileWatcher, ProgressTrackingFileWatcher}
+import de.tu_darmstadt.cs.esa.tapasco.util._
+import de.tu_darmstadt.cs.esa.tapasco.reports._
+import de.tu_darmstadt.cs.esa.tapasco.dse.Heuristics
+import java.nio.file._
+
+import scala.sys.process.{Process, ProcessLogger}
+import scala.util.Properties.{lineSeparator => NL}
+import ComposeResult._
+import LogFormatter._
```
[the remark this answers is not in the document]
Comment (lukasmweber) [reply src/main/scala/tapasco/activity/composers/VivadoComposer.scala]: Noted and corrected.

File: src/main/scala/tapasco/filemgmt/ProgressTrackingFileWatcher.scala (hunk 2 above)
Comment (lukasmweber) [reply src/main/scala/tapasco/filemgmt/ProgressTrackingFileWatcher.scala]: This is now more consistent. `Duration.apply` takes a Long with the duration and a string indicating the unit. The `millis` indicates that the input is in ms.
The `Duration` object is then used to transform into the other units (hours, minutes and seconds).

File: src/main/scala/tapasco/filemgmt/ProgressTrackingFileWatcher.scala (hunk 1 above)
Comment (lukasmweber) [reply src/main/scala/tapasco/filemgmt/ProgressTrackingFileWatcher.scala]: The Watcher is closed automatically when it ends here. In addition it is also closed in `VivadoComposer.scala:89` on exit of the Vivado Process.

File: src/main/scala/tapasco/filemgmt/ProgressTrackingFileWatcher.scala (hunk 3 above)
Comment (lukasmweber) [reply src/main/scala/tapasco/filemgmt/ProgressTrackingFileWatcher.scala]: If I understand the concepts of MultiFileWatcher correctly, the Log is checked every `POLL_INTERVAL` milliseconds. When new lines are added, these are checked in sequence, thus preventing a loss of status.

**Pull request merged**